File: src/main/java/com/blogly/blogly/application/user/signup/SignUpUseCase.java
package com.blogly.blogly.application.user.signup;

import com.blogly.blogly.domain.user.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@RequiredArgsConstructor
@Component
public class SignUpUseCase {

    private final UserRepository userRepository;
    private final PasswordHasher passwordHasher;

    @Transactional
    public SignUpResponse execute(SignUpRequest request) {
        var email = new Email(request.email());

        if (userRepository.existsByEmail(email)) {
            throw new EmailAlreadyExistsException(email.value());
        }

        var user = User.signUp(
                email,
                Password.create(request.password(), passwordHasher),
                new Name(request.name())
        );

        userRepository.save(user);

        return new SignUpResponse(user.getId().value(), user.getEmail().value(), user.getRole());
    }
}

File: src/main/java/com/blogly/blogly/application/auth/SignUpUseCase.kt
package com.blogly.blogly.application.auth

import com.blogly.blogly.application.auth.dto.SignUpRequest
import com.blogly.blogly.application.auth.dto.SignUpResponse
import com.blogly.blogly.application.shared.IdGenerator
import com.blogly.blogly.application.shared.TsidCodec
import com.blogly.blogly.domain.user.*
import com.blogly.blogly.domain.user.exception.EmailAlreadyExistsException
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class SignUpUseCase(
    private val userRepository: UserRepository,
    private val passwordHasher: PasswordHasher,
    private val idGenerator: IdGenerator
) {
    @Transactional
    fun execute(request: SignUpRequest): SignUpResponse {
        val email = Email(request.email)

        if (userRepository.existsByEmail(email)) {
            throw EmailAlreadyExistsException(email.value)
        }

        val user = User(
            id = UserId(idGenerator.generate()),
            email = email,
            password = Password.create(request.password, passwordHasher),
            name = Name(request.name)
        )

        val savedUser = userRepository.save(user)

        return SignUpResponse(
            TsidCodec.encode(savedUser.id.value),
            savedUser.email.value,
            savedUser.role
        )
    }
}
